File: samples/kmp/iosApp/KsDialogsSampleKmp/SampleLoadingRegistration.swift
import KsDialogs
import SampleShared

/// 共有コードのカスタム Loading の ViewModel 型と iOS の View factory の紐付け。
///
/// 共有コードの ViewModel は Swift の ViewModel 契約に準拠しないため、登録は KMP 面 (`kmp`) へ行う。
/// 共有コードで定義した ViewModel のクラスがそのまま登録キーになり、
/// 共有 Presenter からの表示と同じレジストリを引く。
enum SampleLoadingRegistration {
    /// この Sample が使うカスタム Loading を登録する。
    @MainActor
    static func register() {
        // SwiftUI の View をそのまま返す登録。UIView を返す登録と同じ名前で、戻り値の型だけが違う
        Loading.shared.kmp.register(CustomLoadingViewModel.self) { viewModel in
            CustomLoadingCard(viewModel: viewModel)
        }
    }
}

File: samples/kmp/androidApp/src/main/kotlin/jp/kamusoft/ksdialogs/samples/kmp/android/SampleLoadingRegistration.kt
package jp.kamusoft.ksdialogs.samples.kmp.android

import jp.kamusoft.ksdialogs.Loading
import jp.kamusoft.ksdialogs.samples.kmp.CustomLoadingViewModel

/**
 * 共有コードのカスタム Loading の ViewModel 型と Android の View factory の紐付け。
 *
 * View の型は OS ごとに異なるため、登録は Android Native API に対して行う。
 * 共有コードで定義した ViewModel のクラスがそのまま登録キーになり、
 * 共有 Presenter からの表示と同じレジストリを引く。
 */
internal object SampleLoadingRegistration {
    /** この Sample が使うカスタム Loading を登録する。 */
    fun register() {
        // 中身は ViewModel だけを受け取るので、(Context, VM) を受けるコンストラクタの参照をそのまま渡せる
        Loading.instance.registry.register(CustomLoadingViewModel::class, ::CustomLoadingCardView)
    }
}
